scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/CascadingBackend.scala

```diff
@@ -112,266 +133,251 @@ object CascadingBackend {
+            rec(cv.viaHashJoin)
+          case (DebugPipe(p), rec) =>
+            val inner = rec(p)
+            inner.copy(pipe = new Each(inner.pipe, new Debug))
+          case (EmptyTypedPipe, rec) =>
+            // just use an empty iterable pipe.
+            rec(IterablePipe(List.empty[T]))
+          case (fk@FilterKeys(_, _), rec) =>
+            def go[K, V](node: FilterKeys[K, V]): CascadingPipe[(K, V)] = {
+              val rewrite = Filter[(K, V)](node.input, FlatMappedFn.FilterKeysToFilter(node.fn))
+              rec(rewrite)
+            }
+            go(fk)
+          case (f@Filter(_, _), rec) =>
+            // hand holding for type inference
+            def go[T1 <: T](f: Filter[T1]): CascadingPipe[T] =
+              // TODO, maybe it is better to tell cascading we are
+              // doing a filter here: https://github.com/cwensel/cascading/blob/wip-4.0/cascading-core/src/main/java/cascading/operation/Filter.java
+              rec(FlatMapped(f.input, FlatMappedFn.fromFilter(f.fn)))
+
+            go(f)
+          case (f@FlatMapValues(_, _), rec) =>
+            def go[K, V, U](node: FlatMapValues[K, V, U]): CascadingPipe[T] =
+              rec(FlatMapped[(K, V), (K, U)](node.input, FlatMappedFn.FlatMapValuesToFlatMap(node.fn)))
+
+            go(f)
+          case (fm@FlatMapped(_, _), rec) =>
+            // TODO we can optimize a flatmapped input directly and skip some tupleconverters
+            def go[A, B <: T](fm: FlatMapped[A, B]): CascadingPipe[T] = {
+              val CascadingPipe(pipe, initF, fd, conv) = rec(fm.input)
+              val fmpipe = RichPipe(pipe).flatMapTo[A, T](initF -> f0)(fm.fn)(TupleConverter.asSuperConverter(conv), singleSetter)
+              CascadingPipe.single[B](fmpipe, fd)
+            }
 
-    import pipeCache.cacheGet
+            go(fm)
+          case (ForceToDisk(input), rec) =>
+            val cp = rec(input)
+            cp.copy(pipe = RichPipe(cp.pipe).forceToDisk)
+          case (Fork(input), rec) =>
+            // fork doesn't mean anything here since we are already planning each TypedPipe to
+            // something in cascading. Fork is an optimizer level operation
+            rec(input)
+          case (IterablePipe(iter), _) =>
+            val fd = new FlowDef
+            val pipe = IterableSource[T](iter, f0)(singleSetter, singleConverter).read(fd, mode)
+            CascadingPipe.single[T](pipe, fd)
+          case (f@MapValues(_, _), rec) =>
+            def go[K, A, B](fn: MapValues[K, A, B]): CascadingPipe[_ <: (K, B)] =
+              rec(Mapped[(K, A), (K, B)](fn.input, FlatMappedFn.MapValuesToMap(fn.fn)))
+
+            go(f)
+          case (Mapped(input, fn), rec) =>
+            // TODO: a native scalding Map operation would be slightly more efficient here
+            rec(FlatMapped(input, FlatMappedFn.fromMap(fn)))
+
+          case (m@MergedTypedPipe(_, _), rec) =>
+            OptimizationRules.unrollMerge(m) match {
```

Comment:
This seems odd here, maybe a comment as to why its not just in the optimizer flow and we need it here too?

Reply:
the reason is Merge has arity 2, but cascading merge can take a list. I don't *think* cascading is happy if you pass it a merge with no items, and I didn't want to do one with a single merge, but that maybe is a call.

So, we really have to unroll or we will make a giant linked list of Merge nodes, so, this is a case where some transformation is needed.

Or did you just mean the special casing for 0 and 1 item? Again, I *think* that is so cascading won't blow up.

The other optimizations should definitely be in a rule (`x ++ x == x.flatMap { y => List(y, y) }` for instance).

Reply:
So it might be leaky but i'd have thought that maybe in an ideal we would be able to have the invariant for this operator that if we see a MergedTypedPipe post optimization it should always just be 2 pipes. But it does mean we would need to introduce a new node type of a 'many pipe' merge. Which is maybe not worth it then?